### PR TITLE
[DiLoCo] Add Docs for Streaming DiLoCo

### PR DESCRIPTION
**Pull request by Dr-Left**
# Description

Adds comprehensive reference documentation and a production tutorial for **DiLoCo (Distributed Low-Communication)** and **Streaming DiLoCo** training in MaxText.

### Changes Made
1. **DiLoCo Theory & Mathematics Reference ([`docs/reference/core_concepts/diloco.md`](file:///usr/local/google/home/jzuo/maxtext/docs/reference/core_concepts/diloco.md))**:
   - Details the mathematical formulation of bi-level optimization (Inner local AdamW, pseudo-gradient computation, outer Nesterov momentum).
   - Derives operational Arithmetic Intensity ($AI = \frac{\text{FLOPs}}{\text{Bytes}}$) and physical network rooflines demonstrating how DiLoCo scales effective AI by a factor of $H$.
   - Formulates Streaming DiLoCo parameter fragmentation ($P$ fragments), staggered sync scheduling ($\Delta h = \lfloor H/P \rfloor$), asynchronous apply delay ($V$), and soft parameter blending ($\alpha$).
   - Explains the native JAX SPMD + NNX design (`DiLoCoTrainState`, `drjax.program` multi-placement, `FragmentedTreeManipulator`).
   - Clarifies the distinction between current SPMD algorithmic simulation and future MPMD multi-threading hardware overlap for $V$ and $\alpha$.
   - Integrated into `docs/reference/core_concepts.md`.

2. **DiLoCo Pre-training Tutorial ([`docs/tutorials/diloco_pretraining.md`](file:///usr/local/google/home/jzuo/maxtext/docs/tutorials/diloco_pretraining.md))**:
   - Provides a clear comparison between Vanilla DiLoCo (periodic full-model barrier) and Streaming DiLoCo (continuous pipelined fragment communication).
   - Includes production recipes for:
     - **Recipe 1**: Vanilla DiLoCo Multi-Slice Pre-training (Qwen3-8B on 2x `v5p-128`).
     - **Recipe 2**: Streaming DiLoCo Dense Pre-training (Qwen3-8B on 2x `v5p-128` with $H=37, P=37$).
     - **Recipe 3**: Streaming DiLoCo MoE Pre-training (Qwen3-30B-A3B with Grain data loading).
   - Documents multi-replica Orbax checkpointing, automatic resumption, and single-slice parameter bootstrapping.
   - Provides concrete hyperparameter tuning tables and heuristics ($H = P = N_{\text{layers}} + 1$, Outer LR `0.3`–`0.9` inverse scaling rule).
   - Integrated as a standalone tutorial in `docs/tutorials.md`.

3. **Homepage & Index Updates**:
   - Added DiLoCo and Streaming DiLoCo support announcement to `## 🔥 Latest news 🔥` in `README.md` (rendered on `docs/index.md`).
   - Updated pre-training overview references in `README.md`.

### Future Improvements
- In the current JAX SPMD design, `num_communication_overlapping_steps` and `communication_overlapping_alpha` simulate the algorithmic behavior of delayed weight merging without cross-step asynchronous hardware overlap. Future MPMD multi-threading designs will leverage these parameters for true non-blocking hardware compute/communication overlap.

# Tests

- **Sphinx Documentation Build**: Verified clean HTML compilation with zero errors:
  ```bash
  source maxtext_venv/bin/activate
  cd docs && sphinx-build -b html . _build/html
  ```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: docs/tutorials/diloco_pretraining.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+DEVICE_TYPE="v5p-128" \
+NUM_SLICES="2" \
+RUNNAME="vanilla-dlco-8b-01" \
+XPK_WORKLOAD="vanilla-dlco-01" \
+BASE_OUTPUT_DIRECTORY="gs://your-bucket/maxtext-logs" \
+DATASET_PATH="gs://your-bucket/maxtext-datasets" \
+MODEL_NAME="qwen3-8b" \
+STEPS="1000" \
+CHECKPOINT_PERIOD="100" \
+ENABLE_STREAMING_DILOCO="false" \
```
[the remark this answers is not in the document]
Comment (Dr-Left) [reply docs/tutorials/diloco_pretraining.md]: Fixed in new submission.

File: docs/tutorials/diloco_pretraining.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+| `communication_overlapping_alpha` ($\alpha$) | `0.0`                        | Soft parameter blending factor ($\theta_{\text{inner}} \leftarrow \alpha \theta_{\text{inner}} + (1 - \alpha) \theta_{\text{outer}}$). Simulates soft weight interpolation in current SPMD; will be effective alongside $V$ in future MPMD. `0.0` applies direct replacement.                    |
+
+### Practical Tuning Heuristics
+
+- **Synchronize Every Step ($H = P = N_{\text{layers}} + 1$)**:
+  Setting `diloco_sync_period` equal to `num_diloco_fragments` with $P = N_{\text{layers}} + 1$ (e.g., $H=37, P=37$ for 36-layer models like Qwen3-8B, or $H=49, P=49$ for 48-layer models like Qwen3-30B) ensures a steady, constant stream of background communications by syncing 1 fragment on every local step.
+- **Outer Learning Rate Tuning & Inverse Scaling Rule**:
+  Outer LR should be tuned alongside the inner optimizer learning rate. As a core heuristic:
+  \$$\text{Higher Inner Learning Rate} \implies \text{Lower Outer Learning Rate}$$
+  $$\text{Lower Inner Learning Rate} \implies \text{Higher Outer Learning Rate}$\$
```
[the remark this answers is not in the document]
Comment (Dr-Left) [reply docs/tutorials/diloco_pretraining.md]: Fixed in new submission.

File: docs/reference/core_concepts/diloco.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+3. **Global All-Reduce**:
+   \$$\overline{\Delta \theta} = \frac{1}{K} \sum_{k=1}^K \Delta \theta_k$\$
+
+4. **Outer Optimizer Step (Nesterov Momentum)**:
+   \$$v \leftarrow \beta \cdot v + \overline{\Delta \theta}$$
+   $$\theta_{\text{outer}} \leftarrow \theta_{\text{outer}} - \eta_{\text{outer}} \cdot (\overline{\Delta \theta} + \beta \cdot v)$\$
+
+5. **Broadcast**:
+   \$$\theta_{k, 0} \leftarrow \theta_{\text{outer}} \quad \forall k$\$
```
[the remark this answers is not in the document]
Comment (Dr-Left) [reply docs/reference/core_concepts/diloco.md]: Fixed in new submission.

**Review by huytransformer — APPROVED** [review]
bi level optimization!!

**Review by Shuwen-Fang — APPROVED** [review]
